net: make the list of known message types a compile time constant

vasild opened the pull request:
Turn the `std::vector` to `std::array` because it is cheaper and allows us to have the number of the messages as a compile time constant: `ALL_NET_MESSAGE_TYPES.size()` which can be used in future code to build other `std::array`s with that size.

---

This change is part of https://github.com/bitcoin/bitcoin/pull/29418 but it makes sense on its own and would be good to have it, regardless of the fate of https://github.com/bitcoin/bitcoin/pull/29418. Also, if this is merged, that would reduce the size of https://github.com/bitcoin/bitcoin/pull/29418, thus the current standalone PR.

vasild committed:
net: make the list of known message types a compile time constant

Turn the `std::vector` to `std::array` because it is cheaper and
allows us to have the number of the messages as a compile time
constant: `ALL_NET_MESSAGE_TYPES.size()` which can be used in
future code to build other `std::array`s with that size.

epiccurious commented:
Concept ACK 2e312ea909ae6c466007acf791ea0e3356e954cb.

willcl-ark commented on src/protocol.cpp:
```diff
@@ -47,47 +47,6 @@ const char* WTXIDRELAY = "wtxidrelay";
 const char* SENDTXRCNCL = "sendtxrcncl";
 } // namespace NetMsgType
 
-/** All known message types. Keep this in the same order as the list of
```
There is a comment in src/protocol.h#56 which could be updated to reflect the new constant too:

```cpp
/**
 * Bitcoin protocol message types. When adding new message types, don't forget
 * to update allNetMessageTypes in protocol.cpp.
 */
```

vasild replied:
Updated, thanks!

willcl-ark reviewed (approved):
ACK 2e312ea909ae6c466007acf791ea0e3356e954cb

This is a clean tidy-up, and as mentioned useful for #29418 but stands on it's own (as a smalle improvement) too.

Left one req for comment update if touched again.

vasild commented:
`2e312ea909...808fe4ef80`: update comment, see https://github.com/bitcoin/bitcoin/pull/29421#discussion_r1489333827

vasild committed:
protocol: make message types constexpr

As a side effect the names of the constants do not have to be
repeated in `src/protocol.cpp`.

vasild commented on src/net.cpp:
```diff
@@ -3701,7 +3701,7 @@ CNode::CNode(NodeId idIn,
     if (inbound_onion) assert(conn_type_in == ConnectionType::INBOUND);
 
-    for (const std::string &msg : getAllNetMessageTypes())
+    for (const auto& msg : g_all_net_message_types)
```
Done.

vasild commented on src/protocol.h:
```diff
@@ -267,8 +267,44 @@ extern const char* WTXIDRELAY;
-/* Get a vector of all valid message types (see above) */
-const std::vector<std::string>& getAllNetMessageTypes();
+/** All known message types (see above). Keep this in the same order as the list of messages above. */
+static const std::array g_all_net_message_types{
```
Done as a separate commit because it required changing `NetMsgType::*` constants to `constexpr` as well.

A +36 / -77 change, thanks!

vasild replied:
This is relevant: https://stackoverflow.com/questions/45987571/difference-between-constexpr-and-static-constexpr-global-variable

I made the following experiment:

`s.h`:
```cpp
constexpr const char* s{"aabbcc"};
```

`a.cc`:
```cpp
#include "s.h"
#include <stdio.h>
void a()
{
    printf("a(): %p, %s\n", &s, s);
}
```

`b.cc`:
```cpp
#include "s.h"
#include <stdio.h>
void b()
{
    printf("b(): %p, %s\n", &s, s);
}
```
`main.cc`:
```cpp
void a();
void b();
int main(int, char**)
{
    a();
    b();
    return 0;
}
```

```sh
$ clang++-devel -c a.cc -o a.o
$ clang++-devel -c b.cc -o b.o
$ clang++-devel -c main.cc -o main.o
$ clang++-devel a.o b.o main.o -o main
$ ./main
a(): 0x200578, aabbcc
b(): 0x200580, aabbcc
```
Also checked that the `main` executable contains the string `aabbcc` just once, so it is not duplicated. But there are two different pointers pointing to it.

Then I changed the header `s.h` to:
```cpp
inline constexpr const char* s{"aabbcc"};
```

And re-ran:
```sh
$ ./main
a(): 0x200578, aabbcc
b(): 0x200578, aabbcc
```

now it is the same pointer.

So, I added `inline` to the variables involved.

vasild replied:
In our case the variables are inside a namespace. Hinted by https://github.com/bitcoin/bitcoin/pull/29421#discussion_r1503727466 I changed the above to put `s` inside a namespace and got the same result as above. Tried gcc 12, clang 15/17/19.

vasild commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -21,13 +21,12 @@
     ECC_Start();
     SelectParams(ChainType::REGTEST);
-    g_all_messages = getAllNetMessageTypes();
     std::sort(g_all_messages.begin(), g_all_messages.end());
```
> Why the sort?

https://github.com/bitcoin/bitcoin/pull/28165#discussion_r1491029750

You are right that comparing the pointers here is wrong. I added a custom sort function to compare the contents. Thanks!

vasild commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -21,14 +21,17 @@
+    // Sort, so that changes in the order of elements in g_all_net_message_types
+    // do not cause changes in fuzzing for the same fuzz input.
+    std::sort(g_all_messages.begin(), g_all_messages.end(), [](const char* a, const char* b) {
+        return strcmp(a, b) < 0;
```
Just tried that - if I store the constants as `string_view`, e.g. `constexpr std::string_view VERSION{"version"};`, then I get this greeting:

```
test/util/net.cpp:34:9: error: no matching function for call to 'Make'
   34 |         NetMsg::Make(NetMsgType::VERSION,
      |         ^~~~~~~~~~~~
./netmessagemaker.h:14:23: note: candidate function template not viable: no known conversion from 'const std::string_view' (aka 'const basic_string_view<char>') to 'std::string' (aka 'basic_string<char>') for 1st argument
   14 |     CSerializedNetMsg Make(std::string msg_type, Args&&... args)
      |                       ^    ~~~~~~~~~~~~~~~~~~~~
```

vasild replied:
Then another:

```
net.cpp:3705:31: error: no viable overloaded operator[] for type 'mapMsgTypeSize' (aka 'map<basic_string<char>, unsigned long>')
 3705 |         mapRecvBytesPerMsgType[msg] = 0;
      |         ~~~~~~~~~~~~~~~~~~~~~~^~~~
/usr/include/c++/v1/map:1203:18: note: candidate function not viable: no known conversion from 'const value_type' (aka 'const std::string_view') to 'const key_type' (aka 'const std::string') for 1st argument
 1203 |     mapped_type& operator[](const key_type& __k);
      |                  ^          ~~~~~~~~~~~~~~~~~~~
```

It is possible to get over that by changing `mapMsgTypeSize` and `NET_MESSAGE_TYPE_OTHER` to:
```cpp
inline constexpr std::string_view NET_MESSAGE_TYPE_OTHER{"*other*"};
using mapMsgTypeSize = std::map</* message type */ std::string_view, /* total bytes */ uint64_t>;
```
but then there is another error from `rpc/net.cpp` and I think this would spill off the scope of this PR.

vasild replied:
Changed to `std::string` and dropped the `std::sort` modification. Thanks!

Empact commented on src/protocol.h:
```diff
@@ -55,220 +55,255 @@ class CMessageHeader
  * receiving node at the beginning of a connection.
  */
-extern const char* VERSION;
+inline constexpr const char* VERSION{"version"};
```
You can drop `inline` as `constexpr` is implicitly `inline`.
> A static data member (i.e. static class member or namespace-scope variable) declared constexpr on its first declaration is implicitly an inline variable.
https://en.cppreference.com/w/cpp/language/inline

Empact commented on src/protocol.h:
```diff
@@ -55,220 +55,255 @@ class CMessageHeader
-/* Get a vector of all valid message types (see above) */
-const std::vector<std::string>& getAllNetMessageTypes();
+/** All known message types (see above). Keep this in the same order as the list of messages above. */
+inline constexpr std::array g_all_net_message_types{
```
As a constant rather than a variable, seems this should be `ALL_NET_MESSAGE_TYPES`. See developer-notes.md.

vasild replied:
Done

Empact commented:
Concept ACK

vasild commented:
`c7c613c414...fd6a481995`: rename `g_all_net_message_types` to `ALL_NET_MESSAGE_TYPES` as per https://github.com/bitcoin/bitcoin/pull/29421#discussion_r1503734070

vasild commented:
@maflcko good point! Changed from `const char*` to `std::string` in `fd6a481995...546d13274d` (can't have `constexpr` with `std::string`, thus dropped the `constexpr`, it was never the point of this PR).

vasild commented:
`546d13274d...b3efb48673`: remove now unnecessary explicit conversion to `std::string` in `SimulationTest()`.

vasild commented:
@maflcko, this has a NACK from you and I have addressed your concern. If not ACK, it would be nice if you can at least remove your NACK. Thanks for the review so far!

vasild commented:
ping @maflcko

jonatack reviewed (approved):
ACK b3efb486732f3caf8b8a8e9d744e6d20ae4255ef

maflcko reviewed (approved):
Sorry for missing the ping. Now it looks good, because the only change is vector->array and some style fixups.

utACK b3efb486732f3caf8b8a8e9d744e6d20ae4255ef 🎊

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: utACK b3efb486732f3caf8b8a8e9d744e6d20ae4255ef 🎊
SkLePLuDXM+ymjKH3KbyoYmBU/CTns4Bd4K8JycRF2zRDl2t3QMwQrGtZcNnOZ88NGBjgAWuIpsGTMsQzwHKBw==
```

</details>

willcl-ark reviewed (approved):
ACK b3efb486732f3caf8b8a8e9d744e6d20ae4255ef

Based on range diff from previous ACK, all looks good!
`range-diff 2e312ea...b3efb48`

achow101 commented:
ACK b3efb486732f3caf8b8a8e9d744e6d20ae4255ef

laanwj commented on src/protocol.h:
```diff
@@ -55,220 +55,255 @@ class CMessageHeader
  * receiving node at the beginning of a connection.
  */
-extern const char* VERSION;
+inline constexpr const char* VERSION{"version"};
```
Might as well make these `std::string`? Or is there still a good reason that these are `char*`? (it used to be that the message-making functions took char*, but i don't think that's the case anymore)

willcl-ark replied:
IIRC this was left because `constexpr std::string` was not available, but perhaps it is now in c++20?

laanwj replied:
i mean it's defining an array of `std::string` constexpr below it, so i would expect single std::strings to be possible too, i haven't tried it though

vasild replied:
This was tried: https://github.com/bitcoin/bitcoin/pull/29421#discussion_r1491140321

Here is a draft that changes to `std::string_view` the following:

* `NetMsgType::*`
* `ALL_NET_MESSAGE_TYPES`
* `CMessageHeader:pchCommand` (renamed to `CMessageHeader:m_command`)
* `CNetMessage::m_type`
* `V2_MESSAGE_IDS`
* `V2MessageMap::m_map`
* `V2Transport::m_send_type`
and the functions that handle message types.

The scope of this can be reduced by converting the `std::string_view` to `std::string` at some point. The previous code uses `std::string` (or `char*`) everywhere.

https://github.com/vasild/bitcoin/commits/NetMsgType_string_view/
